### DISCUSSION TITLE
unsigned long truncated to int - Expected?

### DISCUSSION THREAD

**Discussion by mhughson** [cc65/cc65#1578] [Q&A] [answered]
Hopefully this is enough information, but I have a line of code like this:

`total_time += (wram_profile_slot_data[i].wram_best_times[i]) * (60) * (60);`

`total_time` is an `unsigned long`. `wram_best_times` are `unsigned chars`.

I ended up with a bug where total_time was wrong, and debugging a bit revealed that it seemed to be operating as a 16-bit number (int) rather than 32-bit (long). Casting the other data to `unsigned long` fixes the bug, but I thought this might be something the compiler could reasonable intuit. Before logging a bug, I wanted to check if this is just expected.

Fix:
`total_time += (unsigned long)(wram_profile_slot_data[i].wram_best_times[i) * (unsigned long)(60) * (unsigned long)(60);`

**Comment by oliverschmidt** [comment, marked as answer]
As far as I know this (unintuitive) behavior is defined this way by the C language specification.

**Comment by groessler** [comment]
On 10/9/21 2:09 AM, Matthew Hughson wrote:
> 
> Fix:
> |total_time += (unsigned 
> long)(wram_profile_slot_data[i].wram_best_times[i) * (unsigned long)(60) 
> * (unsigned long)(60);|
> 

It should be sufficient to just have one of the multiplicands be 
unsigned long, like

total_time += wram_profile_slot_data[i].wram_best_times[i] * 60 * 60ul;

**Comment by greg-king5** [comment]
Oliver is correct.  The C standard explicitly allows compilers to create operations that can overflow -- with no indications!  cc65 isn't smart enough to avoid overflows.  It's the programmer's responsibility to recognize the possibility, and to avoid that problem.

**Comment by mhughson** [comment]
Thanks all! Marking as answered!
